models/accm.py

```diff
@@ -30,14 +31,24 @@ def forward(self, x):
 
 
 class AccmBlock(nn.Module):
-    def __init__(self, input_layer, h_channels, number_of_rsacm):
+    def __init__(self, input_layer, h_channels, number_of_rsacm, args, rsacm_cost):
         super().__init__()
         self.h_channels = h_channels
         self.number_of_rsacm = number_of_rsacm
+        self.args = args
         self.rsacm_list = nn.ModuleList(
             [(RsacmBlock(input_layer, self.h_channels)) for i in range(0, self.number_of_rsacm)])
+        self.rsacm_cost = rsacm_cost
+
+    def initialize_gating_network(self, activation, args):
+        self.gating_network =  Gating_Network(activation, args)
+        self.optimizer_for_gating_network = torch.optim.Adam(self.gating_network.parameters(), args.lr, weight_decay=0)
 
     def forward(self, x):
+        if self.args.train_gating_networks:
+            gating_network_output =  self.gating_network(x)
+            gating_network_output_index = torch.argmax(gating_network_output, dim=1)
+            costs = torch.tensor([self.rsacm_cost*i for i in range(1,self.number_of_rsacm+1)]).cuda()
         x = x.detach() # Detach x make sure that gradient for this layer won't flow to previous layers.

```

Comment:
We are training our gating network in a end-to-end manner, so this detach must not be called for gating network training.